File: frontend/src/features/meetings/utils/websocket.js
export function getWebSocketUrl(meetingId, guestName, guestEmail) {
  const apiUrl = import.meta.env.VITE_API_URL || "http://localhost:8000/api/v1"
  const host = apiUrl.replace(/^https?:\/\//, "").replace("/api/v1", "")
  const protocol = apiUrl.startsWith("https") ? "wss" : "ws"
  const params = new URLSearchParams()
  const token = localStorage.getItem("access_token")
  if (token) {
    params.set("token", token)
  } else if (guestName) {
    params.set("guest_name", guestName)
    if (guestEmail) {
      params.set("guest_email", guestEmail)
    }
  }
  const qs = params.toString()
  return `${protocol}://${host}/ws/meetings/${meetingId}${qs ? "?" + qs : ""}`
}

export class SignalingClient {
  constructor(meetingId, guestName, guestEmail) {
    this.meetingId = meetingId
    this.guestName = guestName || null
    this.guestEmail = guestEmail || null
    this.ws = null
    this.listeners = new Map()
    this.reconnectAttempts = 0
    this.maxReconnectAttempts = 5
    this.reconnectDelay = 2000
    this.intentionalClose = false
    this.connectionId = null
    this._pendingQueue = []
  }

  connect() {
    const url = getWebSocketUrl(this.meetingId, this.guestName, this.guestEmail)
    this.intentionalClose = false
    this.ws = new WebSocket(url)

    this.ws.onopen = () => {
      this.reconnectAttempts = 0
      this._flushQueue()
      this._emit("connected")
    }

    this.ws.onmessage = (event) => {
      try {
        const message = JSON.parse(event.data)
        this._handleMessage(message)
      } catch (e) {
        console.error("Failed to parse WS message:", e)
      }
    }

    this.ws.onclose = (event) => {
      if (this.ws !== event.target && this.ws !== null) {
        return
      }
      this._emit("disconnected", { code: event.code })
      if (this.intentionalClose || event.code === 4000) {
        return
      }
      if (this.reconnectAttempts < this.maxReconnectAttempts) {
        const delay = this.reconnectDelay * Math.pow(2, this.reconnectAttempts)
        this.reconnectAttempts++
        console.log("[Signaling] reconnect in " + delay + "ms (attempt " + this.reconnectAttempts + "/" + this.maxReconnectAttempts + ")")
        setTimeout(() => this.connect(), delay)
      } else {
        console.log("[Signaling] max reconnect attempts reached")
        this._emit("send_error", { type: "reconnect_failed", message: "Could not reconnect to server." })
      }
    }

    this.ws.onerror = () => {
      console.log("[Signaling] websocket error")
    }
  }

  disconnect() {
    this.intentionalClose = true
    this._pendingQueue = []
    if (this.ws) {
      const ws = this.ws
      this.ws = null
      ws.close(1000)
    }
  }

  send(type, targetConnectionId, payload) {
    const message = {
      type,
      target_connection_id: targetConnectionId || null,
      payload: payload || null,
    }

    if (this.ws && this.ws.readyState === WebSocket.OPEN) {
      this.ws.send(JSON.stringify(message))
      return
    }

    if (this.ws && this.ws.readyState === WebSocket.CONNECTING) {
      this._pendingQueue.push(message)
      console.log("[Signaling] queued: " + type + " (readyState=CONNECTING)")
      return
    }

    console.log("[Signaling] send failed: " + type + " (readyState=" + (this.ws ? this.ws.readyState : "null") + ")")
    this._emit("send_error", { type, message: "WebSocket is not connected." })
  }

  _flushQueue() {
    if (this._pendingQueue.length === 0) return
    const queue = this._pendingQueue
    this._pendingQueue = []
    for (const msg of queue) {
      if (this.ws && this.ws.readyState === WebSocket.OPEN) {
        this.ws.send(JSON.stringify(msg))
      }
    }
  }

  on(event, callback) {
    if (!this.listeners.has(event)) {
      this.listeners.set(event, new Set())
    }
    this.listeners.get(event).add(callback)
    return () => this.listeners.get(event)?.delete(callback)
  }

  _emit(event, data) {
    const callbacks = this.listeners.get(event)
    if (callbacks) {
      callbacks.forEach((cb) => {
        try {
          cb(data)
        } catch (e) {
          console.error("WS listener error:", e)
        }
      })
    }
  }

  _handleMessage(message) {
    const { event, data, sender_connection_id, payload } = message

    switch (event) {
      case "participant_joined":
        this._emit("participant_joined", {
          connectionId: data?.connection_id,
          participantId: data?.participant_id,
          participantStatus: data?.participant_status,
          guestName: data?.guest_name,
          userId: data?.user_id,
          type: data?.type,
          isMuted: data?.is_muted,
        })
        break

      case "participant_waiting":
        this._emit("participant_waiting", {
          connectionId: data?.connection_id,
          participantId: data?.participant_id,
          guestName: data?.guest_name,
          userId: data?.user_id,
          type: data?.type,
        })
        break

      case "participant_left":
        this._emit("participant_left", {
          connectionId: data?.connection_id,
          userId: data?.user_id,
        })
        break

      case "participant_admitted":
        this._emit("participant_admitted", {
          connectionId: data?.connection_id,
          participantId: data?.participant_id,
          guestName: data?.guest_name,
          userId: data?.user_id,
          message: data?.message,
        })
        break

      case "participant_removed":
        this._emit("participant_removed", {
          message: data?.message,
        })
        break

      case "participant_rejected":
        this._emit("participant_rejected", {
          message: data?.message,
        })
        break

      case "mute_changed":
        this._emit("mute_changed", {
          participantId: data?.participant_id,
          isMuted: data?.is_muted,
        })
        break

      case "meeting_ended":
        this._emit("meeting_ended", {
          meetingId: data?.meeting_id,
          message: data?.message,
        })
        break

      case "host_left":
        this._emit("host_left", {
          hostId: data?.host_id,
          hostName: data?.host_name,
          timestamp: data?.timestamp,
        })
        break

      case "waiting_room_status":
        this._emit("waiting_room_status", {
          message: data?.message,
        })
        break

      case "status_check":
        this._emit("status_check", {
          status: data?.status,
        })
        break

      case "muted":
        this._emit("muted", {
          message: data?.message,
        })
        break

      case "offer":
        this._emit("offer", {
          senderConnectionId: sender_connection_id,
          sdp: payload?.sdp,
        })
        break

      case "answer":
        this._emit("answer", {
          senderConnectionId: sender_connection_id,
          sdp: payload?.sdp,
        })
        break

      case "ice-candidate":
        this._emit("ice-candidate", {
          senderConnectionId: sender_connection_id,
          candidate: payload?.candidate,
        })
        break

      case "screen_share_requested":
        this._emit("screen_share_requested", {
          participantId: data?.participant_id,
          connectionId: data?.connection_id,
          guestName: data?.guest_name,
          userId: data?.user_id,
        })
        break

      case "screen_share_permission_granted":
        this._emit("screen_share_permission_granted", {
          participantId: data?.participant_id,
          connectionId: data?.connection_id,
          guestName: data?.guest_name,
          userId: data?.user_id,
        })
        break

      case "screen_share_permission_denied":
        this._emit("screen_share_permission_denied", {
          message: data?.message,
        })
        break

      case "screen_share_started":
        this._emit("screen_share_started", {
          participantId: data?.participant_id,
          connectionId: data?.connection_id,
          guestName: data?.guest_name,
          userId: data?.user_id,
        })
        break

      case "screen_share_stopped":
        this._emit("screen_share_stopped", {
          participantId: data?.participant_id,
          connectionId: data?.connection_id,
        })
        break

      case "host_stopped_screen_share":
        this._emit("host_stopped_screen_share", {
          meetingId: data?.meeting_id,
        })
        break

      case "error":
        this._emit("error", {
          message: message?.message,
        })
        break

      default:
        this._emit("message", message)
    }
  }
}

File: backend/app/modules/meetings/websocket.py
import asyncio
from datetime import datetime, timezone
from fastapi import APIRouter, WebSocket, WebSocketDisconnect, Depends
from uuid import UUID
from typing import Optional
import jwt

from app.core.websocket_manager import ws_connection_manager
from app.core.config import settings
from app.core.logger import logger
from app.modules.meetings.dependencies import get_meetings_service
from app.modules.meetings.service import MeetingService
from app.modules.meetings.enums import ParticipantStatus, MeetingStatus
from app.modules.meetings.constants import WSEvent

router = APIRouter(prefix="/ws/meetings", tags=["Live Meeting Realtime Signaling Hub"])

POLL_INTERVAL = 3.0


async def _safe_cleanup(
    room_id: str,
    connection_id: Optional[str],
    meeting_id: UUID,
    user_id: Optional[UUID],
    guest_name: Optional[str],
    guest_email: Optional[str],
    service: MeetingService,
):
    if connection_id is None:
        return

    try:
        ws_connection_manager.disconnect(room_id, connection_id)
    except Exception as exc:
        logger.error("[meeting_id=%s] [participant_id=%s] cleanup failed: disconnect - %s", room_id, connection_id, exc)

    try:
        participant = await service.disconnect_participant_flow(
            meeting_id, user_id=user_id, guest_email=guest_email
        )
    except Exception as exc:
        logger.error("[meeting_id=%s] [participant_id=%s] cleanup failed: disconnect_participant_flow - %s", room_id, connection_id, exc)
        participant = None

    try:
        await ws_connection_manager.broadcast_to_room(room_id, {
            "event": WSEvent.PARTICIPANT_DISCONNECTED,
            "data": {
                "connection_id": connection_id,
                "user_id": str(user_id) if user_id else None,
                "guest_name": guest_name,
            }
        })
    except Exception as exc:
        logger.error("[meeting_id=%s] [participant_id=%s] cleanup failed: broadcast_to_room - %s", room_id, connection_id, exc)

    if user_id:
        try:
            meeting = await service.get_meeting(meeting_id)
            if meeting.host_id == user_id:
                host_name = await service.repo.get_user_name_by_id(user_id) or "Host"
                await ws_connection_manager.broadcast_to_room(room_id, {
                    "event": WSEvent.HOST_LEFT,
                    "data": {
                        "host_id": str(user_id),
                        "host_name": host_name,
                        "timestamp": datetime.now(timezone.utc).isoformat(),
                        "is_temporary": True,
                    }
                })
        except Exception as exc:
            logger.error("[meeting_id=%s] [participant_id=%s] cleanup failed: host_left broadcast - %s", room_id, connection_id, exc)


@router.websocket("/{meeting_id}")
async def meeting_signaling_endpoint(
    websocket: WebSocket,
    meeting_id: UUID,
    token: Optional[str] = None,
    guest_name: Optional[str] = None,
    guest_email: Optional[str] = None,
    service: MeetingService = Depends(get_meetings_service)
):
    # Accept WebSocket first — before any validation
    await websocket.accept()

    room_id = str(meeting_id)
    connection_id: Optional[str] = None

    logger.info("[meeting_id=%s] connection opened", room_id)

    # Extract user_id from JWT token if present
    user_id = None
    if token:
        try:
            payload = jwt.decode(token, settings.JWT_SECRET_KEY, algorithms=["HS256"])
            user_id = UUID(payload["sub"])
        except (jwt.PyJWTError, ValueError):
            pass

    try:
        # Validate meeting and create/find participant
        participant = await service.join_meeting_flow(
            meeting_id=meeting_id, user_id=user_id, guest_name=guest_name, guest_email=guest_email
        )

        # Use participant.id as connection ID to guarantee uniqueness (guests may share names)
        connection_id = str(participant.id)

        logger.info("[meeting_id=%s] [participant_id=%s] connection authenticated", room_id, connection_id)

        # If this is a reconnecting DISCONNECTED participant, cancel the pending force-leave
        if participant.status == ParticipantStatus.ADMITTED:
            await service.reconnect_participant(meeting_id, participant.id)

        # CRITICAL: Commit the DB session here to release the PostgreSQL row-level lock
        # acquired by join_meeting_flow (status update, participant creation, etc.).
        # Without this commit, the WebSocket's long-lived session holds the lock on the
        # `meetings` row for the entire connection duration. Any subsequent HTTP request
        # that updates the same row (e.g. POST /meetings/{id}/end) will block until the
        # WebSocket disconnects — causing a 30s Axios timeout on the frontend.
        await service.repo.db.commit()

        # Register in connection manager (with participant metadata for logging)
        meeting = await service.get_meeting(meeting_id)
        await ws_connection_manager.connect(
            room_id,
            connection_id,
            websocket,
            metadata={
                "participant_id": str(participant.id),
                "user_id": str(user_id) if user_id else None,
                "guest_name": guest_name,
                "is_host": user_id is not None and meeting.host_id == user_id,
            },
        )

        logger.info("[meeting_id=%s] [participant_id=%s] participant created", room_id, connection_id)

        is_waiting = participant.status == ParticipantStatus.WAITING
        is_admitted = participant.status == ParticipantStatus.ADMITTED

        if is_waiting:
            await ws_connection_manager.broadcast_to_room(room_id, {
                "event": WSEvent.PARTICIPANT_WAITING,
                "data": {
                    "participant_id": str(participant.id),
                    "connection_id": connection_id,
                    "guest_name": guest_name,
                    "user_id": str(user_id) if user_id else None,
                    "type": participant.participant_type.value,
                }
            })
            await ws_connection_manager.send_personal_message({
                "event": WSEvent.WAITING_ROOM_STATUS,
                "message": "Waiting for host to admit you..."
            }, websocket)
        else:
            await ws_connection_manager.broadcast_to_room(room_id, {
                "event": WSEvent.PARTICIPANT_JOINED,
                "data": {
                    "participant_id": str(participant.id),
                    "participant_status": participant.status.value,
                    "connection_id": connection_id,
                    "guest_name": guest_name,
                    "user_id": str(user_id) if user_id else None,
                    "type": participant.participant_type.value,
                    "is_muted": participant.is_muted,
                }
            }, exclude_connection_id=connection_id)

            # Send self-info so the (re)connecting participant can set myParticipant
            await ws_connection_manager.send_personal_message({
                "event": WSEvent.PARTICIPANT_JOINED,
                "data": {
                    "participant_id": str(participant.id),
                    "participant_status": participant.status.value,
                    "connection_id": connection_id,
                    "guest_name": guest_name,
                    "user_id": str(user_id) if user_id else None,
                    "type": participant.participant_type.value,
                    "is_muted": participant.is_muted,
                }
            }, websocket)

        last_muted_state = participant.is_muted

        while True:
            current_state = await service.repo.get_participant_by_id(participant.id)
            if current_state is None:
                break

            meeting_check = await service.repo.get_meeting_status(meeting_id)
            if meeting_check in (MeetingStatus.ENDED, MeetingStatus.IDLE, MeetingStatus.CANCELLED):
                await ws_connection_manager.send_personal_message({
                    "event": WSEvent.MEETING_ENDED,
                    "message": "The meeting has ended."
                }, websocket)
                await websocket.close(code=4000, reason="Meeting ended")
                break

            if current_state.status == ParticipantStatus.REMOVED:
                await ws_connection_manager.send_personal_message({
                    "event": WSEvent.PARTICIPANT_REMOVED,
                    "message": "You have been removed from the meeting."
                }, websocket)
                await websocket.close(code=4000, reason="Removed by host")
                break

            if current_state.status == ParticipantStatus.REJECTED:
                await ws_connection_manager.send_personal_message({
                    "event": WSEvent.PARTICIPANT_REJECTED,
                    "message": "You have been rejected from the meeting."
                }, websocket)
                await websocket.close(code=4000, reason="Rejected by host")
                break

            # Detect admission transition (host admitted via REST)
            if not is_admitted and current_state.status == ParticipantStatus.ADMITTED:
                is_admitted = True
                await ws_connection_manager.send_personal_message({
                    "event": WSEvent.PARTICIPANT_ADMITTED,
                    "message": "You have been admitted to the meeting."
                }, websocket)
                await ws_connection_manager.broadcast_to_room(room_id, {
                    "event": WSEvent.PARTICIPANT_ADMITTED,
                    "data": {
                        "participant_id": str(participant.id),
                        "connection_id": connection_id,
                        "guest_name": guest_name,
                        "user_id": str(user_id) if user_id else None,
                    }
                }, exclude_connection_id=connection_id)
                logger.info("[meeting_id=%s] [participant_id=%s] participant admitted", room_id, connection_id)

            # Detect mute state changes initiated by host
            if is_admitted and current_state.is_muted != last_muted_state:
                last_muted_state = current_state.is_muted
                await ws_connection_manager.send_personal_message({
                    "event": WSEvent.MUTE_CHANGED,
                    "data": {
                        "participant_id": str(participant.id),
                        "is_muted": current_state.is_muted,
                    }
                }, websocket)

            # Non-blocking receive with timeout to allow polling
            try:
                raw_data = await asyncio.wait_for(websocket.receive_json(), timeout=POLL_INTERVAL)
            except asyncio.TimeoutError:
                continue

            if not is_admitted:
                if raw_data.get("type") == "check_admitted":
                    await ws_connection_manager.send_personal_message({
                        "event": WSEvent.STATUS_CHECK,
                        "status": current_state.status.value
                    }, websocket)
                continue

            event_type = raw_data.get("type")

            # Block signaling if muted by host
            if current_state.is_muted and event_type in ["offer", "ice-candidate"]:
                await ws_connection_manager.send_personal_message({
                    "event": WSEvent.MUTED,
                    "message": "You are muted by the host."
                }, websocket)
                continue

            # Self-mute/unmute via WebSocket
            if event_type == "self_mute":
                await service.repo.update_participant(current_state, {"is_muted": True})
                last_muted_state = True
                await ws_connection_manager.broadcast_to_room(room_id, {
                    "event": WSEvent.MUTE_CHANGED,
                    "data": {
                        "participant_id": str(participant.id),
                        "is_muted": True,
                    }
                })
                continue

            if event_type == "self_unmute":
                await service.repo.update_participant(current_state, {"is_muted": False})
                last_muted_state = False
                await ws_connection_manager.broadcast_to_room(room_id, {
                    "event": WSEvent.MUTE_CHANGED,
                    "data": {
                        "participant_id": str(participant.id),
                        "is_muted": False,
                    }
                })
                continue

            # Screen share events
            if event_type == "request_screen_share":
                try:
                    await service.request_screen_share(meeting_id, participant.id)
                except Exception as e:
                    await ws_connection_manager.send_personal_message({
                        "event": WSEvent.ERROR,
                        "message": str(e)
                    }, websocket)
                    continue
                await ws_connection_manager.broadcast_to_room(room_id, {
                    "event": WSEvent.SCREEN_SHARE_REQUESTED,
                    "data": {
                        "participant_id": str(participant.id),
                        "connection_id": connection_id,
                        "guest_name": participant.guest_name,
                        "user_id": str(participant.user_id) if participant.user_id else None,
                    }
                })
                continue

            if event_type == "screen_share_started":
                try:
                    await service.start_screen_share(meeting_id, participant.id, user_id=user_id, guest_name=guest_name)
                except Exception as e:
                    await ws_connection_manager.send_personal_message({
                        "event": WSEvent.ERROR,
                        "message": str(e)
                    }, websocket)
                    continue
                await ws_connection_manager.broadcast_to_room(room_id, {
                    "event": WSEvent.SCREEN_SHARE_STARTED,
                    "data": {
                        "participant_id": str(participant.id),
                        "connection_id": connection_id,
                        "guest_name": participant.guest_name,
                        "user_id": str(participant.user_id) if participant.user_id else None,
                    }
                })
                continue

            if event_type == "screen_share_stopped":
                try:
                    await service.stop_screen_share(meeting_id, participant.id)
                except Exception as e:
                    await ws_connection_manager.send_personal_message({
                        "event": WSEvent.ERROR,
                        "message": str(e)
                    }, websocket)
                    continue
                await ws_connection_manager.broadcast_to_room(room_id, {
                    "event": WSEvent.SCREEN_SHARE_STOPPED,
                    "data": {
                        "participant_id": str(participant.id),
                        "connection_id": connection_id,
                    }
                })
                continue

            # Relay WebRTC signaling
            target_peer = raw_data.get("target_connection_id")
            payload = raw_data.get("payload")

            signaling_message = {
                "event": event_type,
                "sender_connection_id": connection_id,
                "payload": payload
            }

            if target_peer:
                target_ws = ws_connection_manager.active_rooms.get(room_id, {}).get(target_peer)
                if target_ws:
                    await ws_connection_manager.send_personal_message(signaling_message, target_ws)
            else:
                await ws_connection_manager.broadcast_to_room(room_id, signaling_message, exclude_connection_id=connection_id)

    except WebSocketDisconnect:
        await _safe_cleanup(room_id, connection_id, meeting_id, user_id, guest_name, guest_email, service)
        logger.info("[meeting_id=%s] [participant_id=%s] participant disconnected", room_id, connection_id)

    except asyncio.CancelledError:
        await _safe_cleanup(room_id, connection_id, meeting_id, user_id, guest_name, guest_email, service)
        logger.info("[meeting_id=%s] [participant_id=%s] cleanup complete", room_id, connection_id)

    except Exception as e:
        await _safe_cleanup(room_id, connection_id, meeting_id, user_id, guest_name, guest_email, service)
        try:
            await websocket.close(code=4000, reason=str(e))
        except Exception:
            pass
        logger.error("[meeting_id=%s] [participant_id=%s] cleanup failed", room_id, connection_id)
